x/stake/keeper.go

```diff
@@ -33,463 +30,647 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
+	// efficiency case:
+	// if already bonded and power increasing only need to update tendermint
+	if powerIncreasing && oldValidator.Status() == sdk.Bonded {
+		bz := k.cdc.MustMarshalBinary(validator.abciValidator(k.cdc))
+		store.Set(GetTendermintUpdatesKey(ownerAddr), bz)
+		return validator
+	}
 
-	// delete from recent and power weighted validator groups if the validator
-	// exists and add validator with zero power to the validator updates
-	if store.Get(GetRecentValidatorKey(address)) == nil {
-		return
+	// efficiency case:
+	// if was unbonded/or is a new validator - and the new power is less than the cliff validator
+	cliffPower := k.getCliffValidatorPower(ctx)
+	if cliffPower != nil &&
+		(!oldFound || (oldFound && oldValidator.Status() == sdk.Unbonded)) &&
+		bytes.Compare(valPower, cliffPower) == -1 { //(valPower < cliffPower
+		return validator
 	}
-	bz, err := k.cdc.MarshalBinary(candidate.validator().abciValidatorZero(k.cdc))
-	if err != nil {
-		panic(err)
+
+	// update the validator set for this validator
+	updatedVal := k.updateBondedValidatorsNew(ctx, store, validator)
+	if updatedVal.Owner != nil { // updates to validator occured  to be updated
+		validator = updatedVal
 	}
-	store.Set(GetAccUpdateValidatorKey(address), bz)
-	store.Delete(GetRecentValidatorKey(address))
+	return validator
 }
 
-//___________________________________________________________________________
-
-// Get the validator set from the candidates. The correct subset is retrieved
-// by iterating through an index of the candidates sorted by power, stored
-// using the ValidatorsKey. Simultaniously the most recent the validator
-// records are updated in store with the RecentValidatorsKey. This store is
-// used to determine if a candidate is a validator without needing to iterate
-// over the subspace as we do in GetValidators
-func (k Keeper) GetValidators(ctx sdk.Context) (validators []Validator) {
-	store := ctx.KVStore(k.storeKey)
+// XXX TODO build in consideration for revoked
+//
+// Update the validator group and kick out any old validators. In addition this
+// function adds (or doesn't add) a validator which has updated its bonded
+// tokens to the validator group. -> this validator is specified through the
+// updatedValidatorAddr term.
+//
+// The correct subset is retrieved by iterating through an index of the
+// validators sorted by power, stored using the ValidatorsByPowerKey. Simultaniously
+// the current validator records are updated in store with the
+// ValidatorsBondedKey. This store is used to determine if a validator is a
+// validator without needing to iterate over the subspace as we do in
+// GetValidators.
+//
+// Optionally also return the validator from a retrieve address if the validator has been bonded
+func (k Keeper) updateBondedValidators(ctx sdk.Context, store sdk.KVStore) {
+	k.updateBondedValidatorsNew(ctx, store, Validator{})
+}
+func (k Keeper) updateBondedValidatorsNew(ctx sdk.Context, store sdk.KVStore,
```

Comment:
Future note - in the usual use case of this function (not `setParams` though), a maximum of one validator can be kicked out. Given that, we could probably optimize that case more by just identifying the single validator we need to kick (which should be the already-tracked "cliff validator").

Reply:
great idea, couldn't resist implementing it :P